onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+#
+
+print("-- backward pass --")
+loss.backward()
+print(f"-- final gradient: {m.grad} --")
+```
+
+Make sure that the forward pass numbers match the output column you calculated
+in Q1, and the final gradient matches the product of the numbers in your final
+column from Q1.
+
+### Q2
+
+Write code to trace the calculation of `m.grad` by adding backward hooks to
+each of the following tensors:
+
+- `m`
+- `mx`
+- `err`
+- `loss`
+
+Use this as an excuse to get familiar with PyTorch internals if you aren't
+already familiar.  I suggest checking out the PyTorch codebase to your machine,
+then browsing via your IDE or `git grep` to find where the `torch.Tensor` class
+is defined, then looking around in there.
+
+What order do the backward hooks print, and what are the gradients at each
+step?  Explain what you see happening.
+
+### Q3
+
+If you rerun your forward pass (don't reinitialize `m`, but start with
+`mx = x * m`), then rerun `loss.backward()`, what effect does it have on your
+`m.grad`?
+
+Fun fact: multiple forward passes and `backward()` calls without zeroing
+gradients in between is exactly how we implement gradient aggregation in
+`PyTorchTrial` (only we don't reuse the same data for both forward passes).
+
+## Optimizers
+
+In the previous exercise, after calculating the gradients, we applied them to
+the model with the following formula:
+
+    m -= gradient * lr
+
+In deep learning frameworks, you don't normally update model weights directly
+like that.  Normally, updating the model is the job of the optimizer.
+
+In PyTorch, the update formula we applied so far is implemented by the SGD
+optimizer:
+
+```python
+# create an SGD optimzer to apply gradients to our model weights (just m):
+opt = torch.optim.SGD([m], lr=0.001)
+
+# since we already called loss.backward() above, we can step our optimizer to
+# apply those gradients, with the configured lr, to our model:
+opt.step()
+print(f"m after step: {m}")
+```
+
+### Q4:
+
+What is the value of `m` after running `opt.step()`?  Can you confirm that it
+matches the gradient update formula we used?
+
+### More Sophisticated Optimizers
+
+The gradient update formula we have used so far is pretty simple.  This simple
+strategy is limited by some drawbacks:
+
+- If you ever get stuck in a local minimum, the gradient may be zero even
+  though there is a much better loss that could be obtained with different
+  model parameters.  SGD will leave you in the local minimum forever.
+
+- It may be unduly affected by bad data, causing you to keep a lower learning
+  rate (and take longer to train) to compensate.
+
+There are many more sophisticated gradient updates in existence.  We're going
+to discuss a simple addition to the basic formula, which is to add "momentum"
+to our gradient calculation.  Momentum means that the gradient we apply at each
+step is a combination of the gradient we calculated in that step and the
+gradient we calculated in the previous step.  The formula becomes:
```

Comment:
More like "the gradient we apply at each step is a combination of the gradient we calculated in that step and a weighted sum of the all the gradients we computed in previous steps" (where more weight is give to more recent gradients).